### DISCUSSION TITLE
Support for AAD App Registration for Graph API usage in v4

### DISCUSSION THREAD

**Discussion by VerlindenEAM** [microsoft-search/pnp-modern-search#3011] [Ideas]
**Is your feature request related to a problem? Please describe.**
For Graph API usage in version V4, we have some Governance rules, (Graph) API permissions are potentially allowed:
1. If via the well-known SharePoint Online Client Extensibility Service Principal: only as isolated
2. If assigned and used through a proprietary, context-dedicated Service Principal (AAD App Registration)

Our company is not in favor of tenant wide API permissions, granted via the generic and well-known SharePoint service principal: SharePoint Online Client Extensibility Azure AD application.
This certainly also applies to delegated permissions: the risk/ vulnerability is precisely that malicious code under identity but without awareness of the logged in user, retrieves permission data via tenant wide allowed API via Graph API and leaks it out.

**Describe the solution you'd like**
2. Indirect authorization via Service Principal that you create and manage yourself as an organization

**Describe alternatives you've considered**
1. SharePoint Online Client Extensibility Service Principal: only as isolated. But iframes are and will be non-ideal for web experience.

**Additional context**

**Comment by wobba** [comment]
@VerlindenEAM I get your point, but as SPFx typically has this as part of the manifest (https://docs.microsoft.com/en-us/sharepoint/dev/spfx/use-aadhttpclient#request-permissions-to-an-azure-ad-application) I suggest you pull the code and change the principal used and pack it up. 

Adding a config setting in the web part itself for this quickly becomes complicated for most people.

And, all permissions via SPFx is delegated by nature - there is no elevation happening. You can always stick to using the SharePoint provider and assume we have no malicious SP REST code running 😉 as allowing general SP REST calls is unsecure as well if you don't do a code review. The templating engine sanitize css/script to prevent random scripts because of this specific reason.

As for iframes, not happening, and makes connecting web parts impossible.

**Comment by VerlindenEAM** [comment]
@wobba could you build in that you have the choice: default with the generic app principal; and if you need then configure the usage to work through company / tenant specific AppPrincipal

**Comment by wobba** [comment]
Unless we get a lot of push for this I believe changing the config and rebuilding is a better approach for now. I'll tag it as a doc improvement point.

**Comment by jagsridharan** [comment]
We had a similar question @wobba does this mean when we approve these permissions, we are approving the permissions (delegated of course) not only for this app but for any apps/scripts that may be created in the future?

**Comment by wobba** [comment]
@jagsridharan That's how SPFx work (unfortunately).

**Comment by jagsridharan** [comment]
Thanks @wobba  for a quick response. My worry is if we request User.Read  or Site.Read as part of this web part and someone can create an App/PowerShell Script in the future and without requesting permissions or admin consent, they can deploy an app to the site collection app catalog and retrieve all User data or all site data? this is the worry the security team have. My assumption is, they won't be able to retrieve say for example all site information unless they have access to the site. Is this correct?

**Comment by wobba** [comment]
The permissions is only applicable to other SPFx solutions - not PowerShell etc. As long as you trust the permissions per SPFx app/ext you install, you should be good from a security point of view. But would be easier to have a separate app for sure.

And, it's all delegated permissions. You can never read anything you don't have access to originally. A user can in the browser today paste in GET queries for /_api/whatever in SP to read data.

**Comment by jagsridharan** [comment]
Hi @wobba  thanks a lot for the clarifications. Much appreciated

**Comment by pandeyup** [comment]
Hi @wobba one question on same what @jagsridharan  has asked. There is one permission like Mail.Read. What exactly this mean. Does user developing SPFX solution can only read his own email not everyone's email. 
Also for future SPFX solutions do we need to follow this route of authentication rather app registration (in terms of taking advantage of avoiding to register a separate app in case this requires delegated permission).  Let me know if I am not clear here)

**Comment by wobba** [comment]
It's all delegated permissions, meaning it allows the the logged in user to read his/her own e-mail. See https://docs.microsoft.com/en-us/sharepoint/dev/spfx/use-aad-tutorial#deploy-the-solution-and-grant-permissions for more information.
